Checkout: A/B test for hiding cart when handling only one product

klimeryk opened the pull request:
The purpose of this A/B test is to hide the cart in the sidebar during checkout when only one product is in it.

/cc @breezyskies for design review
/cc @stephanethomas because he's awesome like that (also, he should be able to point out just how inappropriately I've (ab)used the redux and/or flux stores... or in the wrong components.. or something else - I'm sure this could have been done better, I'm just new to this whole Redux approach)

breezyskies commented:
Tested a CC purchase with the test variaiton and it works as described. 

A few small things: 

- I'm sometimes seeing the sidebar start to load and then disappear (not sure if this is avoidable with the way we assign the test bucket).
- I spotted this display issue on the plans and domains page (looks like other store pages are being set to no sidebar width):

<img width="1020" alt="screen shot 2016-02-17 at 3 48 26 pm" src="https://cloud.githubusercontent.com/assets/3011211/13128889/fe265298-d58d-11e5-9a3c-eb83cf95540a.png">

Also, I'm having second thoughts on not including the coupon code option and would like @mikeshelton1503's opinion on this. Coupon codes are not heavily used, but I couldn't find any specific stats. My initial thought was to solve this design problem if the test goes well, but I also don't want to create support problems here. Thoughts?

mikeshelton1503 commented:
>Coupon codes are not heavily used, but I couldn't find any specific stats. My initial thought was to solve this design problem if the test goes well, but I also don't want to create support problems here. Thoughts?

Moving the coupon code _seems_ like a relatively easy change. If so, I'd vote to go ahead and move it. I know Jetpack is thinking about using coupon codes soon. We can finesse it later but here's how we could display it for now:

![image](https://cloud.githubusercontent.com/assets/12596797/13131769/c09829e6-d5bb-11e5-8a44-3bf301565e32.png)

The field would still toggle based on clicking the "Have a coupon code?" link.

breezyskies commented:
+1 for @mikeshelton1503 solution, assuming we default to collapsed.

breezyskies commented:
@klimeryk I tried adding the coupon section myself, but it causes the payment form to submit (form within a form). Will need some help with that, along with the sidebar issue on plans and domains.

klimeryk commented:
I've fixed the display issue you've reported, @breezyskies.

As for the coupon - I've tried the damn thing not to submit when clicking the `Apply` button in the `CreditCardPaymentBox`'s `form`, but to no avail. Maybe it's the fever I'm experiencing taking its toll, but I've tried everything and I can't figure out why it's not behaving as it should.
First of all, you can't have a `form` inside a `form`, so I changed the `CartCoupon`'s to a `div`. I changed the `Apply` button's type to `button` that should prevent it from submitting the form. Nope.
I've added `event.stopPropagation();` to the callback. Should stop bubbling the event to the form. Nope.

Sorry, I give up for now. I'll try when I'm feeling better, or maybe you can find someone more capable (and healthier :wink:) than me in the meantime :grin: I really would not mind someone else trying, I'm probably missing something obvious and another pair of :eyes: would be welcome.

breezyskies commented:
Tested again and confirmed that the sidebar issue on plans and domains is fixed. :+1: 

I'm not sure who knows the checkout code best -- maybe @drewblaisdell or @scruffian can point you in the right direction?

umurkontaci commented:
@klimeryk This needs a rebase

drewblaisdell commented on client/components/data/cart/index.jsx:
```diff
@@ -17,7 +21,30 @@ function getStateFromStores() {
+			return;
+		}
+
+		this.props.setSection( null, { hasSidebar: ! oneProductToCheckout } );
```
Is there a reason that this logic can't live in `Checkout` (probably in `componentDidMount` and `componentWillReceiveProps`)?

- It feels slightly strange to wrap `CartData`, a component that already wraps `StoreConnection`, with `connect`.
- We wouldn't need the `isCheckout` prop.
- We wouldn't need to worry about setting/removing event listeners on `CartStore`, because `Checkout` receives new props every time cart data changes.

klimeryk commented:
@umurkontaci: thanks for the ping (GitHub really should have a notification for this).

@drewblaisdell: thanks for the tips - I've reworked this PR to incorporate them. I knew I was doing something unnecessarily complicated :laughing: It looks much better now :bow: 

I'll see if I can figure out that damn coupon now...

klimeryk commented:
OK, I've moved the coupon form to: credit card payment form, PayPal payment form and credits payment form. Could I trouble you, my dear designers, to style it exactly how you need it with some CSS magic? :bow: The style for `cart-coupon` has been moved to [`assets/stylesheets/sections/_checkout.scss`](https://github.com/Automattic/wp-calypso/blob/try/hide-cart-with-one-product-only/assets/stylesheets/sections/_checkout.scss).

Now, about the form "resubmitting" when you click the "Apply" button for the coupon. It turns out it's not actually resubmitting - clicking this button results in the cart being updated, updates are fetched from the server, and during that time the form automatically switches to the loading placeholder. You can observe the same behavior if you visit production/wpcalypso, reach the payment form and then try to apply a coupon from the sidebar cart.
So for now I left it as-is. It's definitely not perfect from UX point of view - if we want to move the coupon code permanently, this should be improved. But I'm reluctant to mess around too much in this code, especially given that the check that is triggering this behavior has [a comment about how important it is](https://github.com/Automattic/wp-calypso/blob/try/hide-cart-with-one-product-only/client/my-sites/upgrades/checkout/checkout.jsx#L181) :grin:

breezyskies commented:
Pushed some CSS changes for the coupon code in the checkout form. 

One small note: We should only show it in this location for the `hidden` variation. For the `original`, it should still show in the sidebar.

klimeryk commented:
> One small note: We should only show it in this location for the hidden variation. For the original, it should still show in the sidebar.

That's not a small note :sob: Oh well, back to the drawing board, I mean, `vim`.
I thought that having the coupon form in one place would be good for the test - it would give the same "base" for both tests. I doubt applying the coupon code (or any problems with it) will affect the test in any significant way, but you never know :grin:
I'll rework it so that its location depends on the variation for the test.

klimeryk commented:
PR reworked to take the variation into account when deciding where to show the coupon code form. Also rebased to clean up the history a bit.

breezyskies commented:
Coupon code looks good in both variations :+1:

umurkontaci commented:
Tests are failing

klimeryk commented:
Thanks for the heads-up, @umurkontaci! It turns out, `abtest` was not properly stubbed (despite the stubs being there before my PR). Or rather, those were leftover stubs - I've "updated" them, by moving to `lib` in their respective `test` directories.

umurkontaci commented on client/my-sites/upgrades/cart/cart-coupon.jsx:
```diff
@@ -77,22 +77,26 @@ module.exports = React.createClass({
+			<div className="form">
 				<input type="text" placeholder={ this.translate( 'Enter Coupon Code', { textOnly: true } ) } onChange={ this.handleCouponInput } value={ this.state.couponInputValue } />
-				<button type="submit" className="button">
+				<button type="button" className="button" onClick={ this.applyCoupon }>
```
`<button>` type is `button` by default, no need to specify

klimeryk replied:
Actually, no, [it's `submit`](https://developer.mozilla.org/en/docs/Web/HTML/Element/button) :) I had to change it to `button`, because otherwise it'd submit the form it was placed in (e.g. `CreditCardPaymentBox`).

umurkontaci commented on client/my-sites/upgrades/cart/cart-coupon.jsx:
```diff
@@ -77,22 +77,26 @@ module.exports = React.createClass({
+
 		return (
-			<form onSubmit={ this.applyCoupon }>
+			<div className="form">
```
Why did you move away from `form`?

klimeryk replied:
Because this PR is moving this component (based on the A/B variation, of course) to payment forms (like `client/my-sites/upgrades/checkout/credit-card-payment-box.jsx`) and [the standards prohibit a `form` inside a `form`](http://stackoverflow.com/a/555970/181497).

umurkontaci replied:
Oh I didn't notice it was a formception. All good

klimeryk replied:
> formception

:laughing: 
They should use that in the specification:

> Every form must be enclosed within a FORM element. There can be several forms in a single document, but the formception is not allowed.

umurkontaci commented:
There's no validation on the coupon code entry. I think at the very least we should have an empty field check. 

Don't forget to update the dates of the test :)

One last thing, I'm not sure but do we need to add any extra tracking information somewhere or the ones we currently have is enough for conducting the test?

Otherwise it looks good and works as advertised.

klimeryk commented:
> There's no validation on the coupon code entry. I think at the very least we should have an empty field check.

There was none to begin with :/ I agree that at least some rudimentary validation should be done, but I think it's beyond the scope of this PR. [I've created an issue to track this](https://github.com/Automattic/wp-calypso/issues/3799).

> Don't forget to update the dates of the test :)

Will do :+1:

> One last thing, I'm not sure but do we need to add any extra tracking information somewhere or the ones we currently have is enough for conducting the test?

@breezyskies: ^?

breezyskies commented:
> One last thing, I'm not sure but do we need to add any extra tracking information somewhere or the ones we currently have is enough for conducting the test?

We shouldn't need any additional tracking. We'll use the `calypso_abtest_start` event for the funnel start and the `calypso_checkout_product_purchase` event for the conversion (both already exist).

umurkontaci commented:
> There was none to begin with :/ I agree that at least some rudimentary validation should be done, but I think it's beyond the scope of this PR. I've created an issue to track this.

:+1: 

Looks like I got all the answers I need. :shipit:
